services/wireguard/resources/allocator.go

```diff
@@ -17,29 +17,155 @@
 
 package resources
 
-import "net"
+import (
+	"errors"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+const interfacePrefix = "myst"
 
 // Allocator is mock wireguard resource handler.
 // It will manage lists of network interfaces names, IP addresses and port for endpoints.
-type Allocator struct{}
+type Allocator struct {
+	Ifaces      map[int]struct{}
+	IPAddresses map[int]struct{}
+	Ports       map[int]struct{}
+	mu          sync.Mutex
+}
+
+// NewAllocator creates new resource pool for wireguard connection.
+func NewAllocator() Allocator {
+	return Allocator{
+		Ifaces:      make(map[int]struct{}),
+		IPAddresses: make(map[int]struct{}),
+		Ports:       make(map[int]struct{}),
+	}
+}
+
+// AbandonedInterfaces returns a list of abandoned interfaces that exist in the system,
+// but was not allocated by the Allocator.
+func (a *Allocator) AbandonedInterfaces() ([]net.Interface, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+
+	list := make([]net.Interface, 0)
+	for _, iface := range ifaces {
+		if strings.HasPrefix(iface.Name, interfacePrefix) {
+			ifaceID, err := strconv.Atoi(strings.TrimPrefix(iface.Name, interfacePrefix))
+			if err == nil {
+				if _, ok := a.Ifaces[ifaceID]; !ok {
+					list = append(list, iface)
+				}
+			}
+		}
+	}
+
+	return list, nil
+}
 
 // AllocateInterface provides available name for the wireguard network interface.
-func (a *Allocator) AllocateInterface() string { return "myst0" }
+func (a *Allocator) AllocateInterface() (string, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	for i := 0; i < 255; i++ {
+		if _, ok := a.Ifaces[i]; !ok {
+			a.Ifaces[i] = struct{}{}
+			return fmt.Sprintf("%s%d", interfacePrefix, i), nil
+		}
+	}
+
+	return "", errors.New("no more unused interfaces")
+}
 
 // AllocateIPNet provides available IP address for the wireguard connection.
-func (a *Allocator) AllocateIPNet() net.IPNet {
-	_, subnet, _ := net.ParseCIDR("10.182.47.0/24")
-	return *subnet
+func (a *Allocator) AllocateIPNet() (net.IPNet, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	var s string
+	for i := 0; i < 255; i++ {
+		if _, ok := a.IPAddresses[i]; !ok {
+			a.IPAddresses[i] = struct{}{}
+			s = fmt.Sprintf("10.182.%d.0/24", i)
+			break
+		}
+	}
+
+	_, subnet, err := net.ParseCIDR(s)
+	return *subnet, err
 }
 
 // AllocatePort provides available UDP port for the wireguard endpoint.
-func (a *Allocator) AllocatePort() int { return 52820 }
+func (a *Allocator) AllocatePort() (int, error) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	for i := 52820; i < 53000; i++ {
```

Comment:
We have less range of ports for allocation than IPs. We could relate those with some range constant..

Reply:
Added a const for that.